### ISSUE TITLE
Remote docker support

### ISSUE THREAD

**Issue by alexdesignworks** [lando/lando#647]
I'm working on using lando on CircleCi Docker builder (not machine builder). They are using remote docker.

Currently, changing engine host is not possible since the config is not merged with values provided in `config.yml`.

https://github.com/lando/lando/blob/master/lib/daemon.js#L206

```
|exports.getEngineConfig = function() {

  // Create the default options
  var config = {
    host: '127.0.0.1',
    socketPath: '/var/run/docker.sock'
  };

  // Slight deviation on Windows due to npipe://
  if (process.platform === 'win32') {
    config.socketPath = '//./pipe/docker_engine';
  }

  // Return the config
  return config;

};
```

Question: Will allowing to specify custom host, port and ip, such as `tcp://35.227.77.180:2376` in `config.yml` connect to remotely started docker? Is this the right place for this change? Is providing host as  `tcp://35.227.77.180:2376` enough or Lando does more checks with socket etc?

[UPDATE]
CircleCI provides the following env vars after provisioning remote docker:
```
Allocating a remote Docker Engine

Assigned Docker Engine request id: 5482736
  provisioning: ..
Remote Docker engine created. Using VM 'prealloc-skuvyg9c-c586b8d4-dbb2-4086-b2a0-3051281ee476'
Created container accessible with:
  DOCKER_TLS_VERIFY=1
  DOCKER_HOST=tcp://35.196.9.70:2376
  DOCKER_CERT_PATH=/tmp/docker-certs138100934
  DOCKER_MACHINE_NAME=5482736
```

**Comment by dustinleblanc** [comment]
We're looking at functional testing on CIrcleCI with Docker right now for Lando Core, so this just became quite important

**Comment by alexdesignworks** [comment]
@dustinleblanc I have just tried beta-40 with all new config changes in CircleCI and still cannot get it working.

I'm using `export LANDO_ENGINE_CONFIG='{"host": "35.231.54.63", "port":"2376"}'` to test the connection. The config is now picked-up by lando, but `lando start` still does not see the remote engine.

Direct call to `docker -H tcp://35.231.54.63:2376 info` works correctly, meaning that either whatever check is running in Lando to see if the docker is running does not use all required variables (`isUp()` in `plugins/lando-engine/engine.js` OR more configuration needs to be passed to Lando through environment variables.

Could you please point me in the right direction because whatever Lando documentation says about remote docker is simply not enough.

Thanks

**Comment by alexdesignworks** [comment]
@serundeputy thanks for looking into this. However, the proposed solution does not address the actual issue in this ticket. We still need to allow Lando to run on remote docker. I believe there is still an issue with how Lando handles configuration.

**Comment by dustinleblanc** [comment]
@pirog Im down, was just circling back to look at this

**Comment by pirog** [comment]
@dustinleblanc awesome. let's set up a hangout to look at this together now that i'm coming up for air a bit. i know @alexdesignworks and the salsa team are hot to trot!

**Comment by pirog** [comment]
Alright @dustinleblanc (and probably wroth a ping ar @alexdesignworks as well) i've pushed up what we were able to work through thus far
https://github.com/lando/lando/compare/647/remote-docker-for-real

Def give it a look-see and verify its what we had whilst DOING IT LIVE. To summarize the things we've discovered (and changed) thus far to get things rolling:

1. We need to also strip `COMPOSE_*` envvars from the Lando process so that OTHER THINGS don't interfere with our compose file management
2. After we clear the slate re `DOCKER_*` envvars we need to inspect our `engineConf` and reset relevant envvars so the docker engine and compose binary know where to look. Currently, `DOCKER_HOST`, `DOCKER_TLS_VERIFY` and `DOCKER_CERT_PATH` seem to be the relevant ones.
3. We need to provide better handling for non-standard `docker` and `docker-compose` bin locations on linux (eg if you've installed from source or on a non-desktopy situation) this is already captured over at #935 

More to come!

**Comment by dustinleblanc** [comment]
@pirog good sumarry, I think @alexdesignworks might be able to get away without #3  if the circle config just installs Lando NOT from source

**Comment by dustinleblanc** [comment]
Just checking in on this, @pirog and I spent a good chunk of time debugging this on the CircleCI server. We've got a new branch going with some updates we've made to make this happier:

https://github.com/lando/lando/tree/647/remote-docker-for-real

builds for the forked repo from @alexdesignworks are here:
https://circleci.com/gh/thinktandem/drupal-dev/16

We've gotten a bit further and need to look at the next step that isn't behaving here

**Comment by pirog** [comment]
@dustinleblanc nice, thanks for the update here. I think we've got a few small tweaks to make still so that our engineConfig sets ALL THE THINGS correctly but definitely getting closer.

**Comment by pirog** [comment]
@dustinleblanc ok i've gotten the remote docker part of this to work!

Pushed the lando changes to
https://github.com/lando/lando/compare/647/remote-docker-for-real
(ignore the failing unit test for now)

Pushed our drupal-dev fork changes to
https://github.com/integratedexperts/drupal-dev/compare/feature/lando...thinktandem:feature/lando

So there is good and bad news here:

1. These changes definitely honor remote docker settings set in our `engineConfig`
2. These changes definitely honor #935 
3. I think we should merge the above regardless

Onto the bad:

1. Check out the new fails we are getting in CircleCI
https://circleci.com/gh/thinktandem/drupal-dev/21
2. I think we are getting this because Lando is predicated on the notion of voluming in relevant files like config file, start up scripts, etc. 
3. In a remote docker engine situation our assumed volume mounts dont work since its cross-host. It's easy to verify this by SSHing into a circle build, attaching to a docker container and seeing the files we should be mounting as directories instead of files.

We should talk about all of this but currently im feeling like lando with remote docker engine might not be possible without SERIOUS refactoring

**Comment by dustinleblanc** [comment]
https://github.com/moby/moby/issues/4023 looks related to this issue, still reading it, but yeah, looks like its not a single solution kind of deal

**Comment by alexdesignworks** [comment]
Hey guys! This is amazing! Thank you so much for looking into this thoroughly.

CircleCI indeed has issues with not being able to mount volumes from host to remote docker. But this is okay for our projects since we are putting everything required inside of the containers:
- dependencies are getting installed during Docker image build
- source code is copied into the container just after images are built.

The reason I used installation from source is to be able to modify the code to debug the build. Of course, I would prefer installation from the package if all this works.

I'm a bit confused whether this has been resolved (keeping in mind that volume mounts do not work and it is okay). Could you please ping me once this is ready to be used by installing from the package. Thanks again!

**Comment by pirog** [comment]
@alexdesignworks that all sounds good then. the volume mount part of this was the only thing that we needed to figure out what to do with but if its not a factor on your end then this is pretty close to done. 

What i'm going to do is clean this up so it passes tests and comes with appropriate documentation updates and then roll our beta.45 release later in the day. 

Once that happens i'm going to email you so we can close things out on that end.

**Comment by pirog** [comment]
Alright this is merged into `master` now. Will ping you @alexdesignworks when we've got the actual release out!

**Comment by sminnee** [comment]
FYI I stumbled across this and had the same error that @pirog mentioned above:
```
oci runtime error: container_linux.go:265: starting container process caused "exec: \"/lando-entrypoint.sh\": permission denied"
```

At a guess, this only works if you drop Lando's default images in favour of your own?
